Fix a deadlock when Run() to create a process.

ajeecai opened the pull request:
Child forked will inherit state of global mutex which may be in intermediate taken status by one of threads of then parent, then in child process it tries to get this mutex and is always pending. One example of mutex is malloc_lock, and there are more.

**If this is confirmed as a proper bug fix, I would like choose Contribution option (1): Allowing us to apply the same patch into PacketiX VPN.**

This problem is found because I find that in Linux when vpnclient is connected, the client can't get IP address from vpn server (Is it intentional or bug too? ). Then in commit [#b7f4b78](https://github.com/ajeecai/SoftEtherVPN/commits/b7f4b78) (For Unix like OS, auto launch dhcp client on tap interface), I tried to do dhcp client using Run. It works but after several times of connect/disconnect, I found something like this by ps to list processes:
```
root      3165  0.0  0.0  53968  2008 ?        S<s  14:11   0:00 /home/ajee/github/SoftEtherVPN/bin/vpnclient/vpnclient execsvc
root      3166  0.4  0.2 1099376 18504 ?       S<l  14:11   0:01 /home/ajee/github/SoftEtherVPN/bin/vpnclient/vpnclient execsvc
root      3292  0.0  0.2 1099376 16192 ?       S<   14:12   0:00 /home/ajee/github/SoftEtherVPN/bin/vpnclient/vpnclient execsvc
```
There is one more vpnclient than normal two and it never quits. 
By gdb into the third process and saw this stack trace:
```
0  0x00007f54770bbf4d in __lll_lock_wait () from /lib64/libpthread.so.0
1  0x00007f54770b7d02 in _L_lock_791 () from /lib64/libpthread.so.0
2  0x00007f54770b7c08 in pthread_mutex_lock () from /lib64/libpthread.so.0
3  0x000000000040937d in UnixMemoryAlloc (size=26) at src/Mayaqua/Unix.c:2128
4  0x00000000004fe5c2 in OSMemoryAlloc (size=26) at src/Mayaqua/OS.c:459
5  0x00000000004d7757 in InternalMalloc (size=26) at src/Mayaqua/Memory.c:4173
6  0x00000000004d7342 in MallocEx (size=10, zero_clear_when_free=0) at src/Mayaqua/Memory.c:4014
7  0x00000000004d75c3 in ZeroMallocEx (size=10, zero_clear_when_free=0) at src/Mayaqua/Memory.c:4148
8  0x00000000004d75a4 in ZeroMalloc (size=10) at src/Mayaqua/Memory.c:4144
9  0x00000000004c8497 in UnixCalcStrToUni (str=0x5a68ac "") at src/Mayaqua/Internat.c:988
10 0x00000000004c9ab7 in CalcStrToUni (str=0x5a68ac "") at src/Mayaqua/Internat.c:2013
11 0x00000000004c959b in CopyStrToUni (str=0x5a68ac "") at src/Mayaqua/Internat.c:1662
12 0x000000000050951d in PrintArgs (fmt=0x5a68ac "", args=0x7f546d81a2d8) at src/Mayaqua/Str.c:3010
13 0x0000000000509628 in Print (fmt=0x5a68ac "") at src/Mayaqua/Str.c:3031
14 0x0000000000408220 in UnixRun (filename=0x5afdc3 "dhclient", arg=0x7f546d81a490 "-r -q vpn_vpn", hide=1, wait=0) at src/Mayaqua/Unix.c:1186
15 0x00000000004fe1ab in OSRun (filename=0x5afdc3 "dhclient", arg=0x7f546d81a490 "-r -q vpn_vpn", hide=1, wait=0) at src/Mayaqua/OS.c:317
16 0x00000000004ccb3f in Run (filename=0x5afdc3 "dhclient", arg=0x7f546d81a490 "-r -q vpn_vpn", hide=1, wait=0) at src/Mayaqua/Kernel.c:688
17 0x0000000000490e72 in SessionMain (s=0x7f545c001b20) at src/Cedar/Session.c:715
18 0x000000000046919f in ClientConnect (c=0x7f5448000a20) at src/Cedar/Protocol.c:5422
19 0x0000000000492020 in SessionConnect (s=0x7f545c001b20) at src/Cedar/Session.c:1194
20 0x0000000000492bdb in ClientThread (t=0x7f545c004710, param=0x7f545c001b20) at src/Cedar/Session.c:1533
21 0x00000000004cd269 in ThreadPoolProc (t=0x7f545c003d60, param=0x7f545c003a80) at src/Mayaqua/Kernel.c:923
22 0x0000000000408a5e in UnixDefaultThreadProc (param=0x7f545c004010) at src/Mayaqua/Unix.c:1676
23 0x00007f54770b5dc5 in start_thread () from /lib64/libpthread.so.0
24 0x00007f5475ee0ced in clone () from /lib64/libc.so.6
```

Obviously, This is stuck in the lock. My change is moving all stuff using the mutex lock out of the child process, to the place before forking. With this fix, it works well. 
(**Attention**, make sure UnixCloseIO would not have some Print or Malloc in the future, otherwise also run into trouble). 

Thanks

ajeecai committed:
Fix a deadlock when Run() to create a process. Child forked will inherit
state of global mutex which may be in intermidiate taken status by one of
threads of the parent, then in child process it tries to get this mutex
and is always pending. One example of mutex is malloc_lock, and there are more.

davidebeatrici commented on src/Mayaqua/Unix.c:
```diff
@@ -1172,7 +1174,26 @@ bool UnixRun(char *filename, char *arg, bool hide, bool wait)
 		arg = "";
 	}
-
+	
```
Extra tab.

davidebeatrici commented on src/Mayaqua/Unix.c:
```diff
@@ -1239,7 +1242,6 @@ bool UnixRun(char *filename, char *arg, bool hide, bool wait)
 		return true;
 	}
 }
-
```
Unrelated change.

davidebeatrici commented on src/Mayaqua/Unix.c:
```diff
@@ -1172,7 +1174,26 @@ bool UnixRun(char *filename, char *arg, bool hide, bool wait)
+			args[i] = t->Token[i - 1];
+		}
+	}
+		
```
Two extra tabs (should be an empty line).

davidebeatrici commented on src/Mayaqua/Unix.c:
```diff
@@ -1162,7 +1162,9 @@ bool UnixRunW(wchar_t *filename, wchar_t *arg, bool hide, bool wait)
 bool UnixRun(char *filename, char *arg, bool hide, bool wait)
 {
 	TOKEN_LIST *t;
+	char** args;
```
The asterisk(s) should be attached to the variable name, to maintain consistency with the rest of the source code.

ajeecai replied:
Thank you, I have almost forgot this pull request :) 
Do you need my new pull request for it or do it yourself?  (Not very familiar with how to improve existing pull request on github ).

davidebeatrici replied:
Unfortunately I don't have write access to your branch, as I'm not part of the team.

You can add changes to the pull request by committing them in your `Run_deadlock` branch.

Don't worry about squashing the commits, it can be done when merging the pull request. :)

ajeecai replied:
I have found git commit -amend can help improve the pull request in place, so it is there now with your feedback. Please check it.

Thanks

davidebeatrici replied:
You're welcome, thank you for applying the changes!

ajeecai commented:
I also correct a typo in commit log from "threads of **then** parent" to "threads of **the** parent". Thanks

ajeecai commented:
I also correct a typo in commit log from "threads of **then** parent" to "threads of **the** parent". 
Thanks

davidebeatrici commented:
Thank you for fixing it.

davidebeatrici committed:
vpnweb: add atlthunk.lib to the dependencies

VpnWebControl.obj : error LNK2019: unresolved external symbol "void __stdcall ATL::__FreeStdCallThunk(void *)" (?__FreeStdCallThunk@ATL@@YGXPAX@Z) referenced in function "public: static void __cdecl ATL::_stdcallthunk::operator delete(void *)" (??3_stdcallthunk@ATL@@SAXPAX@Z)

VpnWebControl.obj : error LNK2019: unresolved external symbol "void * __stdcall ATL::__AllocStdCallThunk(void)" (?__AllocStdCallThunk@ATL@@YGPAXXZ) referenced in function "public: static void * __cdecl ATL::_stdcallthunk::operator new(unsigned int)" (??2_stdcallthunk@ATL@@SAPAXI@Z)

davidebeatrici committed:
Add Windows Driver Kit 7.1.0 directories to projects requiring ATL/MFC

davidebeatrici committed:
BuildAll.cmd: correctly start BuildUtil.exe, check if required software exists and exit in case of error

davidebeatrici committed:
README.md: add AppVeyor build status badge

davidebeatrici committed:
AppVeyor: add configuration file

moatazelmasry2 committed:
remove msvc compiling errors (#499)

chipitsine committed:
appveyor: store build artifacts (#500)

davidebeatrici commented:
@moatazelmasry2 I think that you should enable the rebase by default when using `git pull`:
`git config --global pull.rebase true`

moatazelmasry2 commented:
I messed up yesterday's PR specifically by running
`git pull --rebase tidatida-master origin/master`

While his commit remained at the top of the branch, doing a git push -f, pushed everything except his commit and I broke the PR.

I'm hoping with the cherry picking to reduce the damage

davidebeatrici commented:
I would've added his repository as new remote, checked out his branch, pulled from `origin/master` and then force-pushed.

```
git remote add tidatida git@github.com:tidatida/SoftEtherVPN.git
git fetch --all
git checkout -b tidatida-master tidatida/master
git pull --rebase origin master
git push -f
```

moatazelmasry2 commented:
This is exactly what I did last time, really line for line and word for word what you just wrote.
and it broke the PR.

moatazelmasry2 commented:
But I agree that cherry picking is also not the right way to go. I'll try again in future PRs to rebase

moatazelmasry2 commented:
This time I'll do a regular merge of the PR into master, since in my opinion both squash or rebase would be wrong. Feel free to argue otherwise

davidebeatrici commented:
That would be the best way to merge a pull request, in my opinion.

All of the commits are inserted into the target branch as they are, along with a merge commit which indicates that the commits come from a pull request.

moatazelmasry2 commented:
Your great PR has been merged into the master branch. Thank you for your hard work

chengang-hub commented:
Will the stable version fix this? 
i found a  high cpu load ,SoftEther version:[SoftEther VPN Server (Ver 4.38, Build 9760, rtm)]
Can you help us!!! Thank you so much！
https://github.com/SoftEtherVPN/SoftEtherVPN/discussions/1809
